### PR TITLE
Fix #3327: Create app layer mechanism for saving checkpoints.

### LINKED ISSUE #3327: Create app layer mechanism to save exploration checkpoints.
Modify existing app layer classes to save exploration checkpoints.

Target | Date
--|--
**PR creation:** | 2021-07-04
**PR completion:** | 2021-07-08

### PR DESCRIPTION
**Pull request by MaskedCarrot**
## Explanation

Fixes #3327 

This PR is stacked on top of PR #3403

## Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

## Testing the changes made in this PR

1. Checkpointing is disabled by default until a mechanism to resume checkpoints is also implemented. However it can be enabled by changing the value of the variable isCheckpointingEnabled and shouldSavePartialProgresss to true in TopicLessonsFragment.

2. Since size limit imposed on checkpoint database is set to 2MB, therefore the ProgressDatabaseFull dialog won’t be visible until checkpoint database exceeds 2MB in size So to view the ProgressDatabaseFull dialog box, enable checkpointing and then set the size limit to 0 in ExplorationStorageModule.

## Tests

Test | Espresso test result
---| ---
ExplorationActivityTest | Screenshot from 2021-07-06 10-23-43

## Dialog boxes after this PR

This PR introduces the following three dialog boxes: 

1. **StopExplorationDialog**: Visible to the user when the exit an exploration that is saved and size of checkpoint database has not exceeded.
2. **UnsavedProgressDialog**: Visible to the user when the progress is not saved ( this can happen if checkpointing fails or the lesson was previously completed). This dialog box is also visible to the user when they exit a practice session.
3. **ProgressDatabaseFullDialog**: Visible to the user when the progress is saved but the checkpoint database has exceeded the size limit.

### CODE REVIEW

**Comment by anandwana001** [comment]
Assign me back once the UI layer gets reviewed, if there were any UI changes, the test needs to be reviewed again to prevent multiple reviews here.

**Comment by rt4914** [comment]
@aggarwalpulkit596 Can you take first pass at it? If not, feel free to assign back to me. (Not removing my assignment as I will review it tomorrow.)

**Comment by BenHenning** [comment]
Since the base PR is still up in the air, not reviewing this until that's finalized. Please reassign to me at that point.

**Comment by MaskedCarrot** [comment]
The base PR is in review and it should not have any major changes now. So @rt4914 @vinitamurthi @anandwana001  @BenHenning can you please this PR as well.

File: app/src/main/java/org/oppia/android/app/player/exploration/ExplorationActivityPresenter.kt
Comment (aggarwalpulkit596) [inline app/src/main/java/org/oppia/android/app/player/exploration/ExplorationActivityPresenter.kt]: Why do we need to use `!!` complier should handle this since we have already added a null check

**Review by aggarwalpulkit596 — COMMENTED** [review]
LGTM. Just a nit-pick.

File: app/src/main/java/org/oppia/android/app/player/exploration/ExplorationActivityPresenter.kt
Comment (MaskedCarrot) [reply app/src/main/java/org/oppia/android/app/player/exploration/ExplorationActivityPresenter.kt]: `oldestExplorationId`  can be null but `deleteExplorationProgressById` expects it to be not null. `oldestExplorationId` will be null if there was some error while retrieving the oldest checkpoint.

File: domain/src/main/java/org/oppia/android/domain/exploration/ExplorationDataController.kt
Comment (vinitamurthi) [inline domain/src/main/java/org/oppia/android/domain/exploration/ExplorationDataController.kt]: Add `@param` and `@return` annotations (here and below)

File: app/src/sharedTest/java/org/oppia/android/app/player/exploration/ExplorationActivityTest.kt
Comment (anandwana001) [inline app/src/sharedTest/java/org/oppia/android/app/player/exploration/ExplorationActivityTest.kt]: can we update the comment in a bit more understandable way?

File: app/src/sharedTest/java/org/oppia/android/app/player/exploration/ExplorationActivityTest.kt
Comment (anandwana001) [inline app/src/sharedTest/java/org/oppia/android/app/player/exploration/ExplorationActivityTest.kt:1633]: Do we need to update this in all test files? if any?

**Review by anandwana001 — CHANGES_REQUESTED** [review]
nit things

File: domain/src/main/java/org/oppia/android/domain/exploration/ExplorationDataController.kt
Comment (MaskedCarrot) [reply domain/src/main/java/org/oppia/android/domain/exploration/ExplorationDataController.kt]: done

File: app/src/sharedTest/java/org/oppia/android/app/player/exploration/ExplorationActivityTest.kt
Comment (MaskedCarrot) [reply app/src/sharedTest/java/org/oppia/android/app/player/exploration/ExplorationActivityTest.kt:1633]: ` TestExplorationStorageModule` is added to the test file where we need to change the size of the checkpoint database to test a few specific checkpointing features. 

The test files that need this module are:

1. ExplorationActivityTest
2. ExplorationDataControllerTest
3. ExplorationProgressControllerTest
4. ExplorationCheckpointControllerTest

File: app/src/sharedTest/java/org/oppia/android/app/player/exploration/ExplorationActivityTest.kt
Comment (MaskedCarrot) [reply app/src/sharedTest/java/org/oppia/android/app/player/exploration/ExplorationActivityTest.kt]: done

File: app/src/main/java/org/oppia/android/app/player/exploration/ExplorationActivityPresenter.kt
Comment (MaskedCarrot) [reply app/src/main/java/org/oppia/android/app/player/exploration/ExplorationActivityPresenter.kt]: `!!` removing this gives an error.

**Review by anandwana001 — CHANGES_REQUESTED** [review]
Assign me back once approved the UI part, and then I will review the test cases.

File: app/src/main/java/org/oppia/android/app/player/exploration/ExplorationActivityPresenter.kt
Comment (MaskedCarrot) [reply app/src/main/java/org/oppia/android/app/player/exploration/ExplorationActivityPresenter.kt]: Removed `!!`

**Comment by BenHenning** [comment]
@MaskedCarrot please clarify what aspect of this PR you'd like me to review since I don't have any code owners required here.

Also, please fix the failing CI checks that can be fixed (given that this is a chained PR, the base branch check is expected to fail).

**Comment by MaskedCarrot** [comment]
@BenHenning can you please take a look at ExplorationActivityPresenter and ExplorationProgressController

**Comment by MaskedCarrot** [comment]
@rt4914 please take a look.

File: app/src/main/java/org/oppia/android/app/player/exploration/ExplorationActivity.kt
Comment (rt4914) [inline app/src/main/java/org/oppia/android/app/player/exploration/ExplorationActivity.kt]: These consts are getting used only in `Presenter`. Lets shift them there and make it private.

**Review by aggarwalpulkit596 — APPROVED** [review]
LGTM

File: app/src/main/java/org/oppia/android/app/player/exploration/ExplorationActivityPresenter.kt
Comment (rt4914) [inline app/src/main/java/org/oppia/android/app/player/exploration/ExplorationActivityPresenter.kt]: I think this was named considering the `oldest progress` concept in mind but here we can use `explorationId` as it will always be either null or Id of the current exploration.

WDYT?

Same for Title.

File: app/src/main/java/org/oppia/android/app/player/exploration/ExplorationActivityPresenter.kt
Comment (rt4914) [inline app/src/main/java/org/oppia/android/app/player/exploration/ExplorationActivityPresenter.kt:217]: Add `KDoc` to these public functions.

File: app/src/main/java/org/oppia/android/app/player/exploration/ExplorationActivityPresenter.kt
Comment (rt4914) [inline app/src/main/java/org/oppia/android/app/player/exploration/ExplorationActivityPresenter.kt:226]: Add `KDoc` to these public functions.

File: app/src/main/java/org/oppia/android/app/player/exploration/ExplorationActivityPresenter.kt
Comment (rt4914) [inline app/src/main/java/org/oppia/android/app/player/exploration/ExplorationActivityPresenter.kt:278]: Add `KDoc` to these public functions.

File: app/src/main/java/org/oppia/android/app/player/exploration/ExplorationActivityPresenter.kt
Comment (rt4914) [inline app/src/main/java/org/oppia/android/app/player/exploration/ExplorationActivityPresenter.kt]: It should be `ExplorationActivity`

File: app/src/main/java/org/oppia/android/app/player/stopplaying/ProgressDatabaseFullDialogFragment.kt
Comment (rt4914) [inline app/src/main/java/org/oppia/android/app/player/stopplaying/ProgressDatabaseFullDialogFragment.kt]: Make this private

File: app/src/main/java/org/oppia/android/app/testing/ExplorationTestActivityPresenter.kt
Comment (rt4914) [inline app/src/main/java/org/oppia/android/app/testing/ExplorationTestActivityPresenter.kt]: Nit space
`/* backflowScreen= */ null,`

File: app/src/sharedTest/java/org/oppia/android/app/player/exploration/ExplorationActivityTest.kt
Comment (rt4914) [inline app/src/sharedTest/java/org/oppia/android/app/player/exploration/ExplorationActivityTest.kt:131]: Why was this removed? We need this check to make sure the app is accessible by all and tests pass on it.

**Review by rt4914 — CHANGES_REQUESTED** [review]
@MaskedCarrot Please add exact details to PR description about what this PR does, what steps should reviewer do to check this, what is the output, etc.? Considering that the issue also does not contain information anyone who is not fully aware about this project won't understand what's going on.

All following details to your output:
1. Screenshot that all Espresso Test files which have been created/modified in this PR are passing.
2. Screenshot of UI if any.

File: app/src/main/java/org/oppia/android/app/player/exploration/ExplorationActivityPresenter.kt
Comment (MaskedCarrot) [reply app/src/main/java/org/oppia/android/app/player/exploration/ExplorationActivityPresenter.kt]: @rt4914 This will not be the id of the current exploration instead it is the explorationId of the exploration which has the oldest saved progress in the database.

File: app/src/sharedTest/java/org/oppia/android/app/player/exploration/ExplorationActivityTest.kt
Comment (MaskedCarrot) [reply app/src/sharedTest/java/org/oppia/android/app/player/exploration/ExplorationActivityTest.kt:131]: done

File: app/src/main/java/org/oppia/android/app/testing/ExplorationTestActivityPresenter.kt
Comment (MaskedCarrot) [reply app/src/main/java/org/oppia/android/app/testing/ExplorationTestActivityPresenter.kt]: done

File: app/src/main/java/org/oppia/android/app/player/stopplaying/ProgressDatabaseFullDialogFragment.kt
Comment (MaskedCarrot) [reply app/src/main/java/org/oppia/android/app/player/stopplaying/ProgressDatabaseFullDialogFragment.kt]: done

File: app/src/main/java/org/oppia/android/app/player/exploration/ExplorationActivityPresenter.kt
Comment (MaskedCarrot) [reply app/src/main/java/org/oppia/android/app/player/exploration/ExplorationActivityPresenter.kt]: done

File: app/src/main/java/org/oppia/android/app/player/exploration/ExplorationActivityPresenter.kt
Comment (MaskedCarrot) [reply app/src/main/java/org/oppia/android/app/player/exploration/ExplorationActivityPresenter.kt:278]: done

File: app/src/main/java/org/oppia/android/app/player/exploration/ExplorationActivityPresenter.kt
Comment (MaskedCarrot) [reply app/src/main/java/org/oppia/android/app/player/exploration/ExplorationActivityPresenter.kt:226]: done

File: app/src/main/java/org/oppia/android/app/player/exploration/ExplorationActivityPresenter.kt
Comment (MaskedCarrot) [reply app/src/main/java/org/oppia/android/app/player/exploration/ExplorationActivityPresenter.kt:217]: done

File: app/src/main/java/org/oppia/android/app/player/exploration/ExplorationActivityPresenter.kt
Comment (MaskedCarrot) [reply app/src/main/java/org/oppia/android/app/player/exploration/ExplorationActivityPresenter.kt]: renamed this to `oldestCheckpointExplorationId`

File: app/src/main/java/org/oppia/android/app/player/exploration/ExplorationActivity.kt
Comment (MaskedCarrot) [reply app/src/main/java/org/oppia/android/app/player/exploration/ExplorationActivity.kt]: done

**Comment by MaskedCarrot** [comment]
@rt4914 PTAL. I have added the requested changes and also updated the PR description with instructions on how to test out the changes made in this PR.

File: app/src/sharedTest/java/org/oppia/android/app/player/exploration/ExplorationActivityTest.kt
Comment (anandwana001) [inline app/src/sharedTest/java/org/oppia/android/app/player/exploration/ExplorationActivityTest.kt]: Don't use this function, we use testcoroutinedispatchers class and ignore using this wait fucntion

File: app/src/sharedTest/java/org/oppia/android/app/player/exploration/ExplorationActivityTest.kt
Comment (anandwana001) [reply app/src/sharedTest/java/org/oppia/android/app/player/exploration/ExplorationActivityTest.kt:1633]: Are all test in this class passing on robolectric?

**Review by anandwana001 — CHANGES_REQUESTED** [review]

File: app/src/sharedTest/java/org/oppia/android/app/player/exploration/ExplorationActivityTest.kt
Comment (MaskedCarrot) [reply app/src/sharedTest/java/org/oppia/android/app/player/exploration/ExplorationActivityTest.kt:1633]: Yes, all tests added here are passing on robolectric

File: app/src/sharedTest/java/org/oppia/android/app/player/exploration/ExplorationActivityTest.kt
Comment (MaskedCarrot) [reply app/src/sharedTest/java/org/oppia/android/app/player/exploration/ExplorationActivityTest.kt]: This function is not being used by any of the tests added in this PR. `waitForMatch` has a TODO to solve this.

File: app/src/main/java/org/oppia/android/app/player/exploration/ExplorationActivityPresenter.kt
Comment (rt4914) [reply app/src/main/java/org/oppia/android/app/player/exploration/ExplorationActivityPresenter.kt]: Ack.

**Review by rt4914 — APPROVED** [review]
@MaskedCarrot Just make sure that the base branch is develop.

File: app/src/sharedTest/java/org/oppia/android/app/player/exploration/ExplorationActivityTest.kt
Comment (anandwana001) [inline app/src/sharedTest/java/org/oppia/android/app/player/exploration/ExplorationActivityTest.kt]: remove this function as this is not getting used by any function

**Review by anandwana001 — APPROVED** [review]
LGTM

File: app/src/main/java/org/oppia/android/app/player/exploration/ExplorationActivityPresenter.kt
Comment (BenHenning) [inline app/src/main/java/org/oppia/android/app/player/exploration/ExplorationActivityPresenter.kt]: Function KDocs shouldn't start with "This function...". Instead, start with the verb.

Ditto everywhere in this PR.

File: app/src/main/java/org/oppia/android/app/player/stopplaying/StopStatePlayingSessionWithSavedProgressListener.kt
Comment (BenHenning) [inline app/src/main/java/org/oppia/android/app/player/stopplaying/StopStatePlayingSessionWithSavedProgressListener.kt]: The verb tense & capitalization is wrong here & above--please update to be consistent with the broader codebase & our conventions.

File: app/src/main/java/org/oppia/android/app/player/stopplaying/ProgressDatabaseFullDialogFragment.kt
Comment (BenHenning) [inline app/src/main/java/org/oppia/android/app/player/stopplaying/ProgressDatabaseFullDialogFragment.kt]: Please never create strings programmatically--we should always use variable interpolation from strings.xml so that translators can rearrange the string correctly. Ditto anywhere else this might've been done in this PR.

File: app/src/main/res/values/strings.xml
Comment (BenHenning) [inline app/src/main/res/values/strings.xml:44]: This experience seems confusing to me--'Continue' doesn't indicate to the user that they are about to delete progress. What do the PRD & mocks specify in this case?

File: domain/src/main/java/org/oppia/android/domain/exploration/ExplorationProgressController.kt
Comment (BenHenning) [inline domain/src/main/java/org/oppia/android/domain/exploration/ExplorationProgressController.kt]: We shouldn't use exceptions in this way. Exceptions should only be used in exceptional cases. Here, we expect these errors to potentially happen which means returning an enum status would probably be clearer & avoids the potential confusion that these exceptions are catching actual errors vs. legitimate states.

Also, conceptually it's a successful state if the checkpoint can't be saved for some detected reason. AsyncResult's error state should also be exception/reserved for unexpected issues that are detected.

**Review by BenHenning — CHANGES_REQUESTED** [review]
Thanks @MaskedCarrot. Had a few concerns that I think you need to address looking at the specific files that you asked me to. I think my comment from your base PR regarding the DataProvider approach is pertinent here since it can largely simplify how the data flow mechanism is working for exiting the exploration.

If possible, we should try to avoid blocking the exit on a DataProvider since that may result in user-perceived jank.

**Review by vinitamurthi — APPROVED** [review]
LGTM thanks!

File: domain/src/main/java/org/oppia/android/domain/exploration/ExplorationProgressController.kt
Comment (MaskedCarrot) [reply domain/src/main/java/org/oppia/android/domain/exploration/ExplorationProgressController.kt]: changed the approach to dataProvider approach so this is above changes are no longer needed.

File: app/src/main/res/values/strings.xml
Comment (MaskedCarrot) [reply app/src/main/res/values/strings.xml:44]: The mocks say continue. Although they also say click to continue in the dialog description. What should the buttons be here?

Screenshot from 2021-07-15 11-27-51

File: app/src/main/java/org/oppia/android/app/player/stopplaying/StopStatePlayingSessionWithSavedProgressListener.kt
Comment (MaskedCarrot) [reply app/src/main/java/org/oppia/android/app/player/stopplaying/StopStatePlayingSessionWithSavedProgressListener.kt]: done

File: app/src/main/java/org/oppia/android/app/player/exploration/ExplorationActivityPresenter.kt
Comment (MaskedCarrot) [reply app/src/main/java/org/oppia/android/app/player/exploration/ExplorationActivityPresenter.kt]: done

File: app/src/main/java/org/oppia/android/app/player/stopplaying/ProgressDatabaseFullDialogFragment.kt
Comment (MaskedCarrot) [reply app/src/main/java/org/oppia/android/app/player/stopplaying/ProgressDatabaseFullDialogFragment.kt]: changed this to use variable interpolation.

File: app/src/main/res/values/strings.xml
Comment (MaskedCarrot) [reply app/src/main/res/values/strings.xml:44]: And also if I add the line "Click continue to proceed", do I add it in continuation with the dialog description or after leaving a line like shown in the mock?

File: app/src/sharedTest/java/org/oppia/android/app/player/exploration/ExplorationActivityTest.kt
Comment (MaskedCarrot) [reply app/src/sharedTest/java/org/oppia/android/app/player/exploration/ExplorationActivityTest.kt]: done

File: app/src/main/java/org/oppia/android/app/topic/questionplayer/QuestionPlayerActivity.kt
Comment (MaskedCarrot) [inline app/src/main/java/org/oppia/android/app/topic/questionplayer/QuestionPlayerActivity.kt:132]: These are the two methods that have names that can be a bit misleading in QuestionPlayerActivity.

**Review by MaskedCarrot — COMMENTED** [review]
@BenHenning PTAL.

**Review by MaskedCarrot — COMMENTED** [review]
These are the methods whose names can be a bit misleading in questionPlayerActivity

File: app/src/main/java/org/oppia/android/app/topic/questionplayer/QuestionPlayerActivity.kt
Comment (MaskedCarrot) [inline app/src/main/java/org/oppia/android/app/topic/questionplayer/QuestionPlayerActivity.kt]: These names can be a bit misleading.

File: app/src/main/java/org/oppia/android/app/topic/questionplayer/QuestionPlayerActivity.kt
Comment (BenHenning) [inline app/src/main/java/org/oppia/android/app/topic/questionplayer/QuestionPlayerActivity.kt]:
```suggestion
    // This function is not needed because there is no progress that is being saved in training
```

File: app/src/main/java/org/oppia/android/app/topic/questionplayer/QuestionPlayerActivity.kt
Comment (BenHenning) [reply app/src/main/java/org/oppia/android/app/topic/questionplayer/QuestionPlayerActivity.kt]: So I think my concern here is that the training session experience shouldn't even show the flow that results in progress being deleted (i.e. I'd expect the old dialog to show up & neither of these methods to get called).

File: app/src/main/res/values/strings.xml
Comment (BenHenning) [reply app/src/main/res/values/strings.xml:44]: I'm not sure that I understand @MaskedCarrot. What does the current UI look like?

File: scripts/assets/test_file_exemptions.textproto
Comment (BenHenning) [inline scripts/assets/test_file_exemptions.textproto:269]: These actually should have tests to verify that their buttons perform the expected results.

**Review by BenHenning — COMMENTED** [review]
Thanks @MaskedCarrot. I had a few concerns left, but otherwise the PR LGTM.

File: scripts/assets/test_file_exemptions.textproto
Comment (MaskedCarrot) [reply scripts/assets/test_file_exemptions.textproto:269]: There are tests to verify the buttons works correctly in ExplorationActivityTest

File: app/src/main/res/values/strings.xml
Comment (MaskedCarrot) [reply app/src/main/res/values/strings.xml:44]: The current UI looks like this

File: app/src/main/java/org/oppia/android/app/topic/questionplayer/QuestionPlayerActivity.kt
Comment (MaskedCarrot) [reply app/src/main/java/org/oppia/android/app/topic/questionplayer/QuestionPlayerActivity.kt]: The user still sees the old dialog box with every thing is exactly the same as it was from the user's viewpoint.
But the code for this has changed since from explorations progress also has to be deleted. 

Since all the existing dialog boxes have progress associated with them (saved, unsaved and databaseFull), so should I file a new issue that after milestone 1 the training session should have a new dialog box?

File: app/src/main/java/org/oppia/android/app/topic/questionplayer/QuestionPlayerActivity.kt
Comment (MaskedCarrot) [reply app/src/main/java/org/oppia/android/app/topic/questionplayer/QuestionPlayerActivity.kt]: done

File: app/src/main/res/values/strings.xml
Comment (BenHenning) [reply app/src/main/res/values/strings.xml:44]: As discussed over chat, we resolved the confusion & the code has been cleaned up to be a lot clearer.

File: scripts/assets/test_file_exemptions.textproto
Comment (BenHenning) [reply scripts/assets/test_file_exemptions.textproto:269]: As discussed in chat, nearly all files should have tests to verify them in isolation but we're making an exception here for this PR since the dialogs are still being tested via ExplorationActivityTest.

**Review by BenHenning — APPROVED** [review]
Thanks @MaskedCarrot! LGTM for reviewed files.

**Comment by BenHenning** [comment]
Given that past tests have passed, I'm fine with enabling auto-merge here since Gradle tests should be sufficient for the latest changes.

**Pull request merged**
